README.md
# coredumpbot
[![Build Status](https://travis-ci.org/coredump-ch/coredumpbot.svg?branch=master)](https://travis-ci.org/coredump-ch/coredumpbot)
[![Coverage Status](https://coveralls.io/repos/github/coredump-ch/coredumpbot/badge.svg?branch=master)](https://coveralls.io/github/coredump-ch/coredumpbot?branch=master)

Start the Bot like this:

    TELEGRAM_BOT_TOKEN=XXXXXXXXXXXXXXXXXXXXXXX cargo run
 
or use our bot: https://telegram.me/CoredumpBot

![Screenshot0.png](./Screenshot0.png)

# Install on Arch Linux

    sudo pacman -S git cargo
    git clone https://github.com/coredump-ch/coredumpbot.git
    cd coredumpbot
    TELEGRAM_BOT_TOKEN=XXXXXXXXXXXXXXXXXXXXXXX cargo run --release

# Update local Client

    cd coredumpbot
    rm -rf target
    git pull
    TELEGRAM_BOT_TOKEN=XXXXXXXXXXXXXXXXXXXXXXX cargo run --release

# Enable logging

    RUST_LOG=info TELEGRAM_BOT_TOKEN=XXXXXXXXXXXXXXXXXXXXXXX cargo run

# Set commands on Telegram

Use the [@BotFather](https://telegram.me/BotFather) with the command `/setcommands`.

    status - Get People now present
    webcam - Get Picture of Webcam
    location - Get Location of Hackerspace
    subscribe - Subscribe to a Sensor with an Duration
    help - Get Help
    cancel - Cancel Subscription
    version - Get Version of Bot


CHANGELOG.md
# Changelog

This project follows the semantic versioning without the leading v in the tags.

Every Message starts with one of the following words:

 - `Add` for new functionality
 - `Change` for change in behavior
 - `Break` for breaking-change
 - `Fix`
 - `Switch` internal change
 - `Remove` functionality
 - `Update`

## [Unreleased]

## [0.3.2] - 2017-02-12

 - Add systemd service unit `coredumpbot.service`
 - Add new category `Update`
 - Update dependencies

## [0.3.1] - 2016-10-24

 - Fix /help

## [0.3.0] - 2016-08-11

 - Change grammer to be included at compile time
 - Add screenshot

## [0.2.10] - 2016-06-26

 - Fix wording of /status
 - Change answer to /grammar to Markdown

## [0.2.9] - 2016-05-17

 - Remove /crowd
 - Add temperatures and status text to /status

## [0.2.8] - 2016-05-13

 - Add /location

## [0.2.7] - 2016-03-22

 - Add fetch pictures from all webcams
 - Add CHANGELOG.md
 - Implement /webcam {nth}

## [0.2.6] - 2015-11-25

 - Change logging to env_logger crate
 - Change response behavior of bot in groups
 - Add bot logo

## [0.2.5] - 2015-11-23

 - Switch to types from spaceapi crate

## [0.2.4] - 2015-11-18

 - Change /crowd to /status
 - Switch to grammar parsing
 - Add more grammar rules

## [0.2.3] - 2015-11-16

 - Add travis ci integration
 - Add /version
 - Add /help

## [0.2.2] - 2015-11-16

 - Fix telegram-bot crate
 - Add exponential backoff

## [0.2.1] - 2015-11-14

 - Add grammar for user input

## [0.2.0] - 2015-11-12

 - Add /crowd to answer people_now_present

## [0.1.0] - 2015-11-11

 - Initial Release

[Unreleased]: https://github.com/coredump-ch/coredumpbot/compare/v0.3.1...HEAD
[0.3.1]: https://github.com/coredump-ch/coredumpbot/compare/0.3.0...0.3.1
[0.3.0]: https://github.com/coredump-ch/coredumpbot/compare/0.2.10...0.3.0
[0.2.10]: https://github.com/coredump-ch/coredumpbot/compare/0.2.9...0.2.10
[0.2.9]: https://github.com/coredump-ch/coredumpbot/compare/0.2.8...0.2.9
[0.2.8]: https://github.com/coredump-ch/coredumpbot/compare/0.2.7...0.2.8
[0.2.7]: https://github.com/coredump-ch/coredumpbot/compare/0.2.6...0.2.7
[0.2.6]: https://github.com/coredump-ch/coredumpbot/compare/0.2.5...0.2.6
[0.2.5]: https://github.com/coredump-ch/coredumpbot/compare/0.2.4...0.2.5
[0.2.4]: https://github.com/coredump-ch/coredumpbot/compare/0.2.3...0.2.4
[0.2.3]: https://github.com/coredump-ch/coredumpbot/compare/0.2.2...0.2.3
[0.2.2]: https://github.com/coredump-ch/coredumpbot/compare/0.2.1...0.2.2
[0.2.1]: https://github.com/coredump-ch/coredumpbot/compare/0.2.0...0.2.1
[0.2.0]: https://github.com/coredump-ch/coredumpbot/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/coredump-ch/coredumpbot/compare/f6b6d059d86...0.1.0


src/main.rs
//! # CoredumpBot
//! 
//! Checks https://status.crdmp.ch/ and communicates via https://api.telegram.org/ to the user.

extern crate telegram_bot;
extern crate hyper;
extern crate rustc_serialize;
extern crate spaceapi;
extern crate env_logger;
#[macro_use] extern crate log;
extern crate chrono;

use telegram_bot::{Api, ListeningMethod, Message, MessageType, ListeningAction};

pub mod user_input_compiler;
use user_input_compiler::Input;

pub mod spaceapi_client;

pub mod grammar;

use std::time::Duration;

fn main() {
    env_logger::init().unwrap();
    let max_backoff_seconds = Duration::from_secs(128);
    let min_backoff_seconds = Duration::from_secs(1);
    let mut backoff_seconds = min_backoff_seconds;
    let mut sac = spaceapi_client::SpaceApiClient::init();
    let mut last_processed_message_id = 0;
    
    loop {
        // Create bot, test simple API call and print bot information
        let api = Api::from_env("TELEGRAM_BOT_TOKEN").unwrap();
        info!("getMe: {:?}", api.get_me());
        let mut listener = api.listener(ListeningMethod::LongPoll(None));
        
        
        // Fetch new updates via long poll method
        let res = listener.listen(|u| {
            // Restore backoff_seconds, since it works agan
            backoff_seconds = min_backoff_seconds;
            
            
            // If the received update contains a message...
            if let Some(m) = u.message {
                // Discard Messages from Groups the Bot is no longer part of
                if m.message_id == last_processed_message_id {
                    warn!("Dropped Message: {:?}", m);
                    return Ok(ListeningAction::Continue);
                } else {
                    last_processed_message_id = m.message_id;
                }
                
                let name = m.from.first_name;

                // Match message type
                match m.msg {
                    MessageType::Text(t) => {
                        // Print received text message to stdout
                        info!("<{}> {}", name, t);
                        let t = t.replace("@CoreDumpBot", "");
                        let ts:String = format!("{}", t.trim() );
                        
                        match Input::from(ts) {
                        Input::WebCam{ nth } => {
                            let cams = sac.get_webcams();
                            
                            let no_filter = if let Some(nth) = nth {
                                if nth >= cams.len() {
                                    try!(send_message(&api, m.chat.id(),
                                        format!("You requested the webcam #{}, but there are just {}", nth, cams.len())
                                    ));
                                    return Ok(ListeningAction::Continue);
                                }
                                
                                false
                            } else {
                                true
                            };
                            
                            let mut n : usize = 0;
                            let w =
                                cams.iter().filter(|_| {
                                    let b = no_filter || Some(n) == nth;
                                    n += 1;
                                    b
                                })
                            ;
                            
                            for pic_path in w {
                                let caption = sac.basename(&pic_path);
                                match sac.get_tmp_path_for_webcam(&pic_path) {
                                    Ok(pic_tmp_path) => {
                                        try!(api.send_photo(
                                                m.chat.id(),
                                                pic_tmp_path, // Path
                                                Some(caption.into()), // caption
                                                None, // reply_to_message_id
                                                None  // reply_markup
                                        ));
                                    },
                                    Err(e) => {
                                        println!("WebCam({:?}) Error: {:?}",nth, e);
                                        // TODO send Error
                                    },
                                }
                            }
                        },
                        Input::Help => {
                            try!(send_message(&api, m.chat.id(),
                                    "No such help 😜\nuse /webcam for a snapshot of the 3d printer.\nuse /status for an update on people now present\nuse /grammar to receive the spec".into())
                            );
                        },
                        Input::Status => {
                            let s = match sac.fetch_aggregated_status() {
                                Ok(people_now_present) => people_now_present,
                                Err(e) => format!("An error occured 😕\n{}", e),
                            };
                            try!(send_message(&api, m.chat.id(),s));
                        },
                        Input::Start => {
                            try!(send_message(&api, m.chat.id(),
                                    "Welcome to CoredumpBot\nuse /help for a some commands.".into())
                            );
                        },
                        Input::Version => {
                            try!(send_message(&api, m.chat.id(),
                                    format!("Version: {}", env!("CARGO_PKG_VERSION")))
                            );
                        },
                        Input::Grammar => {
                            try!(
                              api.send_message(
                                m.chat.id(),     // chat_id                  : Integer
                                grammar::get_grammar_string(),     // text                     : String
                                Some(telegram_bot::types::ParseMode::Markdown),        // parse_mode               : Option<ParseMode>
                                None,        // disable_web_page_preview : Option<bool>
                                None,        // reply_to_message_id      : Option<Integer>
                                None)        // reply_markup             : Option<ReplyMakrup>
                            );
                        },
                        Input::Location => {
                            let loc = sac.get_location();
                            try!(api.send_location(
                                    m.chat.id(),
                                    loc.lat as f32, loc.lon as f32,
                                    None, None
                            ));
                        },
                        Input::InvalidSyntax( msg ) => {
                            if m.chat.is_user() {
                                try!(send_message(&api, m.chat.id(),
                                        format!("InvalidSyntax: {}\ntry /grammar", msg)
                                ));
                            }
                        },
                        _ => {
                            if m.chat.is_user() {
                                try!(
                                    send_message(&api, m.chat.id(),
                                        "Unknown Command ... try /help".into())
                                );
                            }
                        }, 
                        }
                    },
                    _ => {
                        if m.chat.is_user() {
                            try!(
                                send_message(&api, m.chat.id(),
                                    "Unknown Command ... try /help".into())
                            );
                        }
                    }
                }
            }

            // If none of the "try!" statements returned an error: It's Ok!
            Ok(ListeningAction::Continue)
        });

        if let Err(e) = res {
            warn!("An error occured: {}\nSleeping for {} Seconds", e, backoff_seconds.as_secs());
            // Rest for 10 Seconds
            std::thread::sleep(backoff_seconds);
            
            if backoff_seconds < max_backoff_seconds {
                backoff_seconds = backoff_seconds * 2u32;
            }
        }
    }
}

fn send_message(api: &Api, chat_id: i64, message: String) -> Result<Message,telegram_bot::Error> {
    api.send_message(
        chat_id,     // chat_id                  : Integer
        message,     // text                     : String
        None,        // parse_mode               : Option<ParseMode>
        None,        // disable_web_page_preview : Option<bool>
        None,        // reply_to_message_id      : Option<Integer>
        None)        // reply_markup             : Option<ReplyMakrup>
}



src/grammar.rs
//! Processes the grammar

/// Returns everything from "src/user_input_compiler.rs" util it passes an empty Line.
pub fn get_grammar_string() -> String {
  let file = include_str!("../src/user_input_compiler.rs");
  let mut grammar = String::new();
  let mut max_lines_to_read = 100;
  let mut skipping_header = true;
  
  for line in file.split('\n') {
    if max_lines_to_read > 0 {
      max_lines_to_read -= 1;
    } else {
      break;
    }
    
    if line == "//! ```" {
      skipping_header = false;
    }
    if skipping_header {
      continue;
    }

    if line == "" { // Block Change
      return grammar;
    }

    let (_, line) = line.split_at(4);
    grammar = grammar + line + "\n";
  }
  
  grammar
}


#[cfg(test)]
mod test {
  use super::*;
  
  #[test]
  fn print_grammar_string() {
    let g :String = get_grammar_string();
    println!("alles: {:?}", g);
    // TODO real Test
    assert!(g.is_empty() == false);
    assert_eq!(format!(":= *\n```\n"), g[(g.len()-9)..]);
  }
}


src/grammer.rs
use std::io;
use std::io::prelude::*;
use std::io::BufReader;
use std::fs::File;

/// Returns everything from "src/user_input_compiler.rs" util it passes an empty Line.
pub fn get_grammer_string() -> String {
  let f = match File::open("src/user_input_compiler.rs") {
    Ok(f) => { println!("grammer::File::open"); f },
    Err(e) => return format!("grammer::open error: {:?}", e),
  };
  let mut reader = BufReader::new(f);
  let mut grammer = String::new();
  let mut i = 100;
  
  loop {
    let mut buffer = String::new();
    // read a line into buffer
    match reader.read_line(&mut buffer) {
      Ok(v) => { println!("grammer::read_line: {:?}", v); },
      Err(e) => return format!("grammer::read_line error: {:?}", e),
    }
    println!("buffer: {:?}", buffer);
    grammer = grammer + &*buffer;
    if buffer == "\n" {
      return grammer;
    }
    
    if i > 0 {
      i -= 1;
    } else {
      break;
    }
  }
  
  grammer
}


#[cfg(test)]
mod test {
  use super::*;
  
  #[test]
  fn print_grammer_string() {
    let g :String = get_grammer_string();
    println!("alles: {:?}", g);
    // TODO real Test
    assert!(g.is_empty() == false);
    //assert!("\n\n".is_suffix_of(g));
  }
}


src/cached_value.rs
//! Handle information which expires

use std::time::{Instant, Duration};
use std::ops::*;

/// A fully generic container for expireing information
#[derive(Debug)]
pub struct CachedValue<T> {
  data : T,
  expires : Instant,
  lifetime : Duration,
}

impl<T> CachedValue<T> {
  pub fn new(data: T, lifetime: Duration) -> CachedValue<T> {
    CachedValue {
      data: data,
      expires: Instant::now() + lifetime,
      lifetime: lifetime,
    }
  }

  pub fn has_expired(&self) -> bool {
    self.expires < Instant::now()
  }
}

impl<T> Deref for CachedValue<T> {
  type Target = T;

  fn deref(&self) -> &T {
    &self.data
  }
}

/// Using this will renew the expiration:
impl<T> DerefMut for CachedValue<T> {
  fn deref_mut(&mut self) -> &mut T {
    self.expires = Instant::now() + self.lifetime;
    &mut self.data
  }
}


pub trait CacheContainer {
  fn remove_expired(&mut self);
}

impl<T> CacheContainer for ::std::vec::Vec<CachedValue<T>> {
  fn remove_expired(&mut self) {
    // Maybe switch to swap_remove()
    self.retain(|&ref e| e.has_expired() == false);
  }
}
impl<T> CacheContainer for ::std::collections::VecDeque<CachedValue<T>> {
  fn remove_expired(&mut self) {
    // Maybe switch to swap_remove()
    self.retain(|&ref e| e.has_expired() == false);
  }
}




#[cfg(test)]
mod test {
  use super::*;
  //use super::CacheContainer;
  use std::time::{Instant, Duration};
  use std::thread;
  use std::collections::VecDeque;

  #[test]
  fn expired_cache() {
    let c = CachedValue::new(42, Duration::new(0,0));

    sleep(1);

    assert_eq!(true, c.has_expired());
  }

  #[test]
  fn deref() {
    let c = CachedValue::new(42, Duration::new(0,0));

    assert_eq!(42, *c);
  }

  #[test]
  fn deref_mut() {
    let mut c = CachedValue::new(42, Duration::new(0,0));

    let expires_0 = c.expires;

    sleep(42);
    *c = 23;

    let expires_1 = c.expires;

    assert_eq!(23, *c);
    assert!(expires_0 < expires_1);
  }

  #[test]
  fn use_with_vec() {
    let mut v = vec![
      CachedValue::new(21, Duration::from_millis(1)),
      CachedValue::new(42, Duration::from_millis(4)),
    ];

    sleep(2);
    v.remove_expired();

    assert_eq!(1, v.len());
    assert_eq!(42, *v[0]);
  }

  #[test]
  fn use_with_vec_deque() {
    let mut v : VecDeque<_> = vec![
      CachedValue::new(21, Duration::from_millis(1)),
      CachedValue::new(42, Duration::from_millis(42)),
    ].into_iter().collect();

    sleep(2);
    v.remove_expired();

    assert_eq!(1, v.len());
    assert_eq!(42, *v[0]);
  }

  fn sleep(ms : u64) {
    thread::sleep( Duration::from_millis(ms) );
  }
}


src/user_input_compiler.rs
//! Process user input

//! # user_input_compiler
//!
//! Parses Input with this Grammar after trimming the Input and ignoring Whitespaces:
//! 
//! ```
//! Command         := "/" CommandWord
//! CommandWord     := Status | Subscribe | Cancel | Version | Help | WebCam | Start | Grammar | Location | InvalidSyntax
//! Status          := "status"
//! Subscribe       := "subscribe" SensorSelector Duration
//! SensorSelector  := SensorString OptionalInteger
//! SensorString    := "account_balance" | "barometer" | "beverage_supply" | "door_locked" | "humidity" | "network_connections" | "power_consumption" | "temperature" | "total_member_count" | "radiation.alpha" | "radiation.beta_gamma" | "radiation.beta" | "radiation.gamma" | "people_now_present" | "wind"
//! OptionalInteger := Integer | ɛ
//! Integer         := [0-9]*
//! Duration        := Real TimeSuffix
//! TimeSuffix      := "m" | "min" | "h" | "d"
//! Real            := Integer "." Integer | Integer
//! Cancel          := "cancel"
//! Version         := "version"
//! Help            := "help"
//! WebCam          := "webcam" OptionalInteger
//! Start           := "start"
//! Grammar         := "grammar"
//! Location        := "location"
//! InvalidSyntax   := *
//! ```

// ===========================================================================

use std::time::Duration;
use std::str::Chars;

#[derive(Debug, PartialEq)]
pub enum Input {
  Status,
  Subscribe{ sensor :SensorSelector, duration :Duration },
  Cancel,
  Version,
  Help,
  WebCam{ nth :Option<usize> },
  Start,
  Grammar,
  Location,
  InvalidSyntax( String ),
}
#[derive(Debug, PartialEq)]
pub struct SensorSelector {
  sensor_selector :String,
  nth :Option<u64>,
}
use self::Input::*;

/// like try! but unwraps the Error
macro_rules! extract {
    ($expr:expr) => (match $expr {
        Result::Ok(val) => val,
        Result::Err(err) => {
            return err
        }
    })
}

/// Use like this:
///
/// ```
/// assert_eq!( Help, Input::from( format!("/help") ) )
/// ```
impl From<String> for Input {
  /// Start the Parser/Compiler
  fn from(s :String) -> Input {
    let s = s.trim();
    
    if s.len() == 0 {
      return InvalidSyntax( format!("Empty Request") );
    }
    
    let mut s = s.chars();
    
    if s.next().unwrap() != '/' {
      return InvalidSyntax( format!("Command must start with /") );
    }
    
    match_command_word(&mut s)
  }
}

fn match_command_word(s :&mut Chars) -> Input {
  if starts_with(s, "status") {
    return Status;
  } else
  if matches_with(s, "subscribe") {
    let sensor = extract!(match_sensor_selector(s));
    let duration = extract!(match_duration(s));
    return Subscribe{ sensor: sensor, duration: duration };
  } else
  
  if starts_with(s, "cancel") {
    return Cancel;
  } else
  
  if starts_with(s, "version") {
    return Version;
  } else
  
  if starts_with(s, "help") {
    return Help;
  } else
  
  if matches_with(s, "webcam") {
    let nth = match match_full_integer(s) {
      Ok(n) if n >= 0 => Some(n as usize),
      Ok(_) => return InvalidSyntax("Expected positive Integer".into()),
      Err(_) => None,
    };
    return WebCam{ nth: nth };
  } else
  
  if starts_with(s, "grammar") {
    return Grammar;
  } else 
  
  if starts_with(s, "location") {
    return Location;
  } else 
  
  if starts_with(s, "start") {
    return Start;
  } else {
  
    return InvalidSyntax( "Invalid CommandWord".into() );
  }
}

fn match_sensor_selector(s :&mut Chars) -> Result<SensorSelector,Input> {
  let sensor;
  
  let w = consume_whitespaces(s);
  
  if starts_with(s, "account_balance") {
    sensor = "account_balance";
  } else
  if starts_with(s, "barometer") {
    sensor = "barometer";
  } else
  if starts_with(s, "beverage_supply") {
    sensor = "beverage_supply";
  } else
  if starts_with(s, "door_locked") {
    sensor = "door_locked";
  } else
  if starts_with(s, "humidity") {
    sensor = "humidity";
  } else
  if starts_with(s, "network_connections") {
    sensor = "network_connections";
  } else
  if starts_with(s, "power_consumption") {
    sensor = "power_consumption";
  } else
  if starts_with(s, "temperature") {
    sensor = "temperature";
  } else
  if starts_with(s, "total_member_count") {
    sensor = "total_member_count";
  } else
  if starts_with(s, "radiation.alpha") {
    sensor = "radiation.alpha";
  } else
  if starts_with(s, "radiation.beta_gamma") {
    sensor = "radiation.beta_gamma";
  } else
  if starts_with(s, "radiation.beta") {
    sensor = "radiation.beta";
  } else
  if starts_with(s, "radiation.gamma") {
    sensor = "radiation.gamma";
  } else
  if starts_with(s, "people_now_present") {
    sensor = "people_now_present";
  } else
  if starts_with(s, "wind") {
    sensor = "wind";
  } else {
    return Err( InvalidSyntax( format!("Invalid SensorSelector: {}", collect_iterator(s)) ) );
  }
  
  debug!("sensor: {}", sensor);
  s.skip(sensor.len() -1 +w).next();
  
  // OptionalInteger
  let mut it = s.clone();
  let mut nth = None;
  if let Ok(n) = match_integer(&mut it) {
    debug!("potential OptionalInteger: {}", n);
    match it.next() {
      Some(ws) => {
        debug!("ws: '{}'", ws);
        if ws == ' ' || ws == '\t' || ws == '\r' || ws == '\n' {
          nth = match match_integer(s) {
            Ok(n) if n >= 0 => {
              debug!("next Integer: {}", n);
              Some(n as u64)
            },
            Ok(n) => return Err( InvalidSyntax( format!("Index {} must be positive", n) ) ),
            Err(_) => None,
          };
        }
      },
      None => {
        nth = None;
      },
    }
  }
  
  
  
  Ok( SensorSelector{ sensor_selector: format!("{}", sensor), nth: nth } )
}

/// Duration        := Real TimeSuffix
fn match_duration(s :&mut Chars) -> Result<Duration,Input> {
  let real :f64 = try!(match_real(s));
  let ts   :i64 = try!(match_timesuffix(s));
  
  let duration = Duration::from_secs((real * (ts as f64)) as u64);
  
  Ok( duration )
}

fn match_real(s: &mut Chars) -> Result<f64, Input> {
  let st :String = try!(collect_real(s));
  st.parse::<f64>().map_err(|msg| InvalidSyntax(format!("Invalid Real: {:?}", msg)) )
}

/// Real            := Integer "." Integer | Integer
fn collect_real(s: &mut Chars) -> Result<String, Input> {
  let mut i1 = try!( collect_integer(s) );
  
  let mut it = s.clone();
  
  if let Some(punkt) = it.next() {
    if punkt != '.' {
      return Ok( i1 );
    }
    s.next(); // consume '.'
    let i2 = try!(collect_integer(s));
    
    i1 = i1 + "." + &i2;
  }
  
  Ok( i1 )
}

fn match_full_integer(s :&mut Chars) -> Result<i64, Input> {
  let _ = consume_whitespaces(s);
  let sign = if let Some(c) = s.clone().next() {
    if c == '-' { -1 } else { 1 }
  } else { 1 };
  
  // Skip '-'
  if sign < 0 {
    s.next();
  }
  
  
  
  let st :String = try!(collect_integer(s));
  match st.parse::<i64>() {
    Ok(val)  => Ok(sign * val),
    Err(msg) => Err( InvalidSyntax(format!("Invalid Integer: {:?}", msg)) ),
  }
}

fn match_integer(s :&mut Chars) -> Result<i64, Input> {
  let st :String = try!(collect_integer(s));
  st.parse::<i64>().map_err( |msg| InvalidSyntax(format!("Invalid Integer: {:?}", msg)) )
}

/// Integer         := [0-9]*
fn collect_integer(s :&mut Chars) -> Result<String, Input> {
  let mut i = format!("");
  let mut it = s.clone();
  
  
  let w = consume_whitespaces(&mut it);
  
  for c in it {
    info!("collect_integer: '{}'", c);
    match c {
      '0' ... '9' => i.push( c ),
      _ => break,
    }
  }
  
  if i.len() == 0 {
    Err( InvalidSyntax(format!("Invalid Integer")) )
  } else {
    s.skip( i.len() -1 +w).next();
    Ok( i )
  }
}

/// TimeSuffix      := "m" | "min" | "h" | "d"
/// Factor to multiply with Seconds
fn match_timesuffix(s :&mut Chars) -> Result<i64, Input> {
  consume_whitespaces(s);
  
  if starts_with(s, "m") || starts_with(s, "min") {
    return Ok(60);
  }
  if starts_with(s, "h") {
    return Ok(60*60);
  }
  if starts_with(s, "d") {
    return Ok(60*60*24);
  }
  
  Err( InvalidSyntax(format!("Invalid TimeSuffix")) )
}


/// Search without modifing the Iterator
fn starts_with(haystack_iter :&Chars, needle :&str) -> bool {
  let mut iter = haystack_iter.clone();
  
  for c in needle.chars() {
    if c != iter.next().unwrap_or('/') {
      return false;
    }
  }
  
  true
}

/// Search and advance the Iterator
fn matches_with(haystack_iter :&mut Chars, needle :&str) -> bool {
  let mut steps_taken = 0;
  let mut iter = haystack_iter.clone();
  
  for c in needle.chars() {
    steps_taken += 1;
    if c != iter.next().unwrap_or('/') {
      return false;
    }
  }
  
  haystack_iter.skip(steps_taken -1).next();
  
  true
}

fn collect_iterator(it :&mut Chars) -> String {
  let mut s = String::new();
  
  for c in it {
    s.push(c);
  }
  
  s
}

fn consume_whitespaces(it :&mut Chars) -> usize {
  let dry_run = it.clone();
  let mut skip :usize = 0;
  
  for c in dry_run {
    match c {
      ' ' | '\t' | '\r' | '\n' => { skip += 1 },
      _ => break,
    }
  }
  
  if skip > 0 {
    it.skip(skip -1).next();
  }
  skip
}




#[cfg(test)]
mod test {
  #![allow(non_snake_case)] // may change to non_snake_case_functions
  use super::*;
  use super::Input::*;
  use super::{starts_with, matches_with, consume_whitespaces, match_duration, match_integer, match_full_integer, match_real, match_timesuffix};
  use std::time::Duration;
  
  
  // =================== Util Tests ===================
  #[test]
  fn starts_with_one() {
    let mut s = "abcdef".chars();
    assert!( starts_with(&mut s, "abcd") );
  }
  
  #[test]
  fn starts_with_two() {
    let mut s = "abcdef".chars();
    assert!( starts_with(&mut s, "xyz") == false );
    assert!( starts_with(&mut s, "abcd") );
  }
  
  #[test]
  fn starts_with_four() {
    let mut s = "abcdef".chars();
    assert!( starts_with(&mut s, "abcxxx") == false );
    assert!( starts_with(&mut s, "abcxyz") == false );
    assert!( starts_with(&mut s, "abcxyz") == false );
    assert!( starts_with(&mut s, "abcd") );
  }
  
  #[test]
  fn starts_with_final_iterator_state() {
    let mut s = "abcdef".chars();
    assert!( starts_with(&mut s, "abcxxx") == false );
    assert!( starts_with(&mut s, "abcd") );
    assert_eq!( Some('a'), s.next() );
  }
  
  #[test]
  fn matches_with_final_iterator_state() {
    let mut s = "abcdef".chars();
    assert!( matches_with(&mut s, "abcxxx") == false );
    assert!( matches_with(&mut s, "abcd") );
    assert_eq!( Some('e'), s.next() );
  }
  
  #[test]
  fn consume_whitespaces_iter_position() {
    let mut s = "    -".chars();
    let w = consume_whitespaces(&mut s);
    
    assert_eq!( 4, w );
    assert_eq!( Some('-'), s.next() );
  }
  
  // =================== Parser Tests ===================
  
  #[test]
  fn empty_2_fail() {
    assert_eq!(InvalidSyntax("Empty Request".into()), Input::from(format!("")))
  }
  
  #[test]
  fn status() {
    assert_eq!(Status, Input::from( format!("/status") ) )
  }
  
  #[test]
  fn crowd() { // cancelled with 0.3.0
    assert_eq!(InvalidSyntax("Invalid CommandWord".into()), Input::from( format!("/crowd") ) )
  }
  
  #[test]
  fn subscribe_pnp_10min() {
    assert_eq!(Subscribe{ sensor: SensorSelector{ sensor_selector: "people_now_present".into(), nth: None }, duration: Duration::from_secs(10*60) } 
        , Input::from( format!("/subscribe people_now_present 10min") ) )
  }
  
  #[test]
  fn subscribe_pnp_13_10min() {
    assert_eq!(Subscribe{ sensor: SensorSelector{ sensor_selector: "people_now_present".into(), nth: Some(13) }, duration: Duration::from_secs(10*60) } 
        , Input::from( format!("/subscribe people_now_present 13 10min") ) )
  }
  
  #[test]
  fn subscribe_pnp_2h() {
    assert_eq!(Subscribe{ sensor: SensorSelector{ sensor_selector: "people_now_present".into(), nth: None }, duration: Duration::from_secs(2*60*60) } 
        , Input::from( format!("/subscribe people_now_present 2h") ) )
  }
  
  #[test]
  fn subscribe_pnp_7d() {
    assert_eq!(Subscribe{ sensor: SensorSelector{ sensor_selector: "people_now_present".into(), nth: None }, duration: Duration::from_secs(7*60*60*24) } 
        , Input::from( format!("/subscribe people_now_present 7d") ) )
  }
  
  #[test]
  fn cancel() {
    assert_eq!(Cancel, Input::from( format!("/cancel") ) )
  }
  
  
  
  #[test]
  fn integer42() {
    let mut s = "42".chars();
    assert_eq!(Ok(42), match_integer(&mut s));
  }
  
  #[test]
  fn integer_neg42() {
    let mut s = "-42".chars();
    assert_eq!(Err(InvalidSyntax("Invalid Integer".into())), match_integer(&mut s))
  }
  
  #[test]
  fn integer__neg42() {
    assert_eq!( Err(InvalidSyntax("Invalid Integer".into())), match_integer(&mut " -42 ".chars()) );
  }
  
  #[test]
  fn real6_6() {
    assert_eq!( Ok(6.6), match_real(&mut "6.6".chars()) )
  }
  
  #[test]
  fn real_123_456_() {
    assert_eq!( Ok(123.456), match_real(&mut " 123.456 ".chars()) )
  }
  
  #[test]
  fn real666_666() {
    assert_eq!( Ok(666.666), match_real(&mut "666.666".chars()) )
  }
  
  #[test]
  fn real10min() {
    assert_eq!( Ok(10 as f64), match_real(&mut "10min".chars()) )
  }
  
  #[test]
  fn real_duration_10min() {
    let mut s = " 10min".chars();
    
    assert_eq!( Ok(10 as f64), match_real(&mut s) );
    assert_eq!( Ok(60), match_timesuffix(&mut s) );
  }
  
  #[test]
  fn real_duration_10__min() {
    let mut s = " 10  min".chars();
    
    assert_eq!( Ok(10 as f64), match_real(&mut s) );
    assert_eq!( Ok(60), match_timesuffix(&mut s) );
  }
  
  #[test]
  fn real_duration__10_5min__() {
    let mut s = "  10.5min  ".chars();
    
    assert_eq!( Ok(10.5), match_real(&mut s) );
    assert_eq!( Ok(60), match_timesuffix(&mut s) );
  }
  
  #[test]
  fn real_duration__10_5__min__() {
    let mut s = "  10.5  min  ".chars();
    
    assert_eq!( Ok(10.5), match_real(&mut s) );
    assert_eq!( Ok(60), match_timesuffix(&mut s) );
  }
  
  #[test]
  fn real12_3__45_6() {
    let mut s = "12.3  45.6".chars();
    
    assert_eq!( Ok(12.3), match_real(&mut s) );
    assert_eq!( Ok(45.6), match_real(&mut s) );
  }
  
  #[test]
  //#[should_panic(expected = "InvalidSyntax(\"Invalid Integer\")")]
  fn real6_punkt_6() {
    assert_eq!( Err(InvalidSyntax("Invalid Integer".into())), match_real(&mut "6..6".chars()) )
  }
  
  
  #[test]
  fn help() {
    assert_eq!( Help, Input::from( format!("/help") ) )
  }
  
  #[test]
  fn version() {
    assert_eq!( Version, Input::from( format!("/version") ) )
  }
  
  #[test]
  fn start() {
    assert_eq!( Start, Input::from( format!("/start") ) )
  }
  
  
  
  #[test]
  fn duration_10min() {
    assert_eq!( Ok(Duration::from_secs(10*60)), match_duration(&mut "10min".chars()) )
  }
  
  
  
  #[test]
  fn match_integer_position() {
    let mut s = "  10  22".chars();
    assert_eq!( Ok(10), match_integer(&mut s) ); // I do not care here
    
    assert_eq!( "  22", s.collect::<String>() );
  }
  
  #[test]
  fn match_integer_position_spaces() {
    let mut s = "  10  ".chars();
    assert_eq!( Ok(10), match_integer(&mut s) ); // I do not care here
    
    assert_eq!( "  ", s.collect::<String>() );
  }
  
  #[test]
  fn match_integer_fail_position_spaces() {
    let mut s = "  bla  ".chars();
    assert_eq!( Err(InvalidSyntax("Invalid Integer".into())), match_integer(&mut s) ); // I do not care here
    
    assert_eq!("  bla  ", s.collect::<String>());
  }
  
  #[test]
  fn match_integer_fail_neg42() {
    assert_eq!( Err( InvalidSyntax("Invalid Integer".into()) ), match_integer(&mut "  -42  ".chars()) )
  }
  
  #[test]
  fn match_full_integer_neg42() {
    assert_eq!( Ok(-42), match_full_integer(&mut "  -42  ".chars()) )
  }
  
  
  
  #[test]
  fn webcam() {
    assert_eq!( WebCam{ nth: None }, Input::from( format!("/webcam") ) )
  }
  
  #[test]
  fn webcam_negative_1() {
    assert_eq!( InvalidSyntax("Expected positive Integer".into()), Input::from( format!("/webcam -1") ))
  }
  
  #[test]
  fn webcam_negative_13() {
    assert_eq!( InvalidSyntax("Expected positive Integer".into()), Input::from( format!("/webcam -13") ))
  }
  
  #[test]
  fn webcam_42() {
    assert_eq!( WebCam{ nth: Some(42) }, Input::from( format!("/webcam 42") ) )
  }
  
  #[test]
  fn webcam_23() {
    assert_eq!( WebCam{ nth: Some(23) }, Input::from( format!("/webcam 23") ) )
  }
  
  
  
  #[test]
  fn grammar() {
    assert_eq!( Grammar, Input::from( format!("/grammar") ) )
  }
  
  
  #[test]
  fn location() {
    assert_eq!( Location, Input::from( format!("/location") ) )
  }

  #[test]
  fn crap_input() {
    assert_eq!( InvalidSyntax("Command must start with /".into()), Input::from( format!("not a valid command!") ))
  }
}
